### PR TITLE
feat: skip redundant adaptation commands

### LINKED ISSUE #589: Redundant adaptation commands not always skipped
This bug has been identified in the discussion of #588. This block in `_adapt_light` is supposed to skip adaptation commands if the current state of a light already equals a desired state (called `service_data`):
https://github.com/basnijholt/adaptive-lighting/blob/9caf64509349958a340f8cef9207fdb0ea35dbef/custom_components/adaptive_lighting/switch.py#L1176-L1186

Unfortunately, this does not always work and can be reproduced by
1. turning on a light,
2. awaiting adaptation,
3. turning off the light,
4. turning the light back on and noticing (e.g. in Z2M logs) that an adjustment command is sent even though the light is already in the desired state (since it was just adapted in step 2).

I can't post meaningful AL logs, since this issue is only noticeable through the absence of the `Cancelling adapt to light` log messages that signal that an adaptation is skipped. In this example, adding a log for `last_service_data` to the block cited above would print `None`. There are more situations in which it happens, and looking at the implementation, the issues seem to be:
- `last_service_data` is only updated by `_adapt_light()`, hence only by adaptation commands from AL
- `last_service_data` is deleted by `reset()`, e.g., when lights are turned off, the configuration is updated, or manual control is changed
- the comparison includes the transition time

In summary, every initial turn-on of a light, and every turn-on call (e.g., to adjust unrelated state) after a regular adaptation command (which sets the transition to an unusual duration, e.g., the default 45 seconds) potentially lead to unnecessary adaptation commands. It happens because the state isn't tracked over the complete lifetime of the lights, and transitions times interfere with state comparisons. 

This might seem like a minor issue, but unnecessary messages in an active Zigbee network can negatively affect it, and also some lights do not handle multiple parallel messages very nicely, resulting in a suboptimal user experience. For example, transitions of Tradfri lights might get jittery or skipped, or the lights even end up in an invalid state (on, but reported off). It gets twice as worse with `separate_turn_on_commands` enabled.

As a solution, I think that the `last_service_data` should be updated by any changes, no matter if from AL or from elsewhere. What's important is that we know the current state of a light to be able to decide whether an adaptation command is necessary, and that's independent of who set it. This could be done in `state_changed_event_listener`. Also, it should never be deleted, because the lights always have a state. With `separate_turn_on_commands` enabled, it should only send the subset of command that actually changes state (e.g., skip the brightness command but send the color command). I implemented a proof of concept and it works nicely.

### PR DESCRIPTION
**Pull request by protyposis**
Reduce adaptation commands to only those that are actually required to reach a light's desired state. For example, if a light's brightness is already at 100% and AL wants to adapt to 100%, the command won't be sent. Analogous for color and the combination of both. Commands are also minimized to the relevant attributes when parts of the desired state are already set, e.g., brightness already matches but color needs to be adapted. Transition times are ignored because they are irrelevant when no change happens.

Builds upon #598 and fixes #589 by replacing the fix for #80 (originally added in #460) with a generic approach. Additional implications:
- decreases the likelihood of #410
- reduced network load in certain situations, e.g., with a high adaptation update frequency as suggested in #614, or many lights switched/adapted at the same time
- improved turn-on responsiveness with `split_adaptation_commands` where lights directly fade to the target color when the target brightness is already set

This feature is inactive by default and can be activated by setting `experimental_skip_redundant_commands`. The config property is prefixed with `experimental_` so that the default can be easily reset for all users (by removing the prefix) in a later version when we have gathered experience from other users (I can only test with Ikea Tradfri and Philips Hue lights via Zigbee2MQTT).

I have put all functionality that can be expressed as pure functions into a separate file to keep the impact on the convoluted `switch.py` minimal and ensure maximal testability. This feature mostly solves edge cases, but it also prepares the architecture for a PR for #594.

### CODE REVIEW

**Comment by protyposis** [comment]
All tests work fine locally in my dev container and Docker test container. @basnijholt, any ideas what could be wrong?

File: custom_components/adaptive_lighting/switch.py [hunk 1]
```diff
@@ -1358,7 +1314,7 @@ async def _update_manual_control_and_maybe_adapt(
                 else:
                     _fire_manual_control_event(self, light, context)
             else:
-                await self._adapt_light(light, transition, force=force, context=context)
+                await self._adapt_light(light, transition, context=context)
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:1317]: Why is the `force` argument no longer passed through here?

Comment (protyposis) [reply custom_components/adaptive_lighting/switch.py:1317]: `force` controlled whether `_adapt_light` should send adaptation commands when the known entity state equals the requested state, and that is now decided by `experimental_skip_redundant_commands`.

I couldn't find out why `force` was enforced in particular situations and why it would be necessary, e.g. https://github.com/basnijholt/adaptive-lighting/blob/36c656b7958f6a9a637b50800534749038e146ee/custom_components/adaptive_lighting/switch.py#L527-L537
and I was hoping you'd know :)

Comment (basnijholt) [reply custom_components/adaptive_lighting/switch.py:1316]: We pass `force=True` when toggling sleep mode for example.

Comment (basnijholt) [reply custom_components/adaptive_lighting/switch.py:1316]: Between all the changes made by @th3w1zard1 and my old code, I can also no longer figure out why that was needed 😅

**Comment by basnijholt** [comment]
Thanks a lot again!

I have removed the "Experimental" marker. It would likely be a breaking change once we remove the experimental from `experimental_skip_redundant_commands` and folks have that in their configs.

**Comment by basnijholt** [comment]
Sorry for the delay BTW. I just became father to 2 babies :tada: 😄
